### PR TITLE
Validations and test

### PR DESCRIPTION
**Pull request by herokudev**
- Use validations for models.
- Write unit tests for models.
- Secure app from n+1 problems.
- Add flash messages.

### CODE REVIEW

**Review by mke2111 — CHANGES_REQUESTED** [review]
### Hey @herokudev, 

**This is a great project you have going on, well-done lad**. 

   
**However, there are a few sections to improve before your project is approved.** 🔴 
<hr>

## Required Changes ♻️
_Check the inline comments under the review for sections to improve._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -23,19 +23,13 @@ def new
...
+    @post = @user.posts.create(params.require(:post).permit(:title, :text))
     if @post.save
-      flash[:success] = 'Post saved successfully'
+      flash[:notice] = 'Post saved successfully'
       redirect_to user_post_url
     else
       flash.now[:error] = 'Error: Post could not be saved'
       render :new, locals: { post: @post }
     end
   end
```
Comment (mke2111) [inline app/controllers/posts_controller.rb:34]: - [ ] In line 4, within the post#index action, you should solve the N+1 problem when fetching all the posts and their comments and likes for a user. Read about the `includes` method in ruby when solving the N+1 problem.

File: spec/models/comment_spec.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                     posts_counter: 0)
+    @user.save
+    @post = Post.new(title: 'Post 1', text: 'This is a post', author_id: 1)
+    @post.save
+  end
+
+  it 'should be equal to zero if there are no recent posts' do
+    recent = @user.recent_posts
+    expect(recent.length).to eq(0)
+  end
```
Comment (mke2111) [inline spec/models/comment_spec.rb:19]: - [ ] This test case is failing. See the screengrab below and give this method a quick fix.
Screenshot from 2021-12-17 03-10-10

File: spec/models/comment_spec.rb [hunk 3]
```diff
@@ -1,6 +1,11 @@
 require 'rails_helper'
 
 RSpec.describe User, type: :model do
+  it 'should be equal to zero if there are no recent posts' do
```
Comment (rloterh) [inline spec/models/comment_spec.rb:4]: This section requires further improvement:
- [ ] A test case is failing here. Kindly ensure to improve upon this aspect.

**Failing test case:**
image

Comment (rloterh) [inline spec/models/comment_spec.rb:4]: - [ ] The issue regarding post controller is not properly resolved, on line 4 of your `posts_controller.rb`. You can improve this aspect by doing similar:
```
 @posts = @user.posts.includes(:comments) 
```

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Herbert, 

You did great on this project, and for making the changes requested in the previous review(s)! 👏👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 🚦  🚦  

## Good Points 👍  
- No linter errors

## Required Changes ♻️
_Kindly check the comments below. Please take your time to carefully improve aspects outlined in the previous code review(s), and please check to see that the changes implemented are working as intended._


<br>
Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have 1 more limited review per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Dipeshtwis — APPROVED** [review]
# Status: Approved ✔️

## Hi @herokudev,

Good job on implementing the suggested changes. Well Done.
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

**You may close this Pull Request and merge the changes**

  
**Keep being awesome devs and good luck with your next milestone!**
---------

Cheers and Happy coding!👏👏👏
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
_**Dipesh Kumar**_

--------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project(check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**